### ISSUE TITLE
Calendar and contact infrastructure

### ISSUE THREAD

**Issue by matrss** [nix-community/home-manager#2335]
### Description

I wanted to setup vdirsyncer today and stumbled across the old PRs for calendar and contact infrastructure and associated with that the vdirsyncer and khal modules in #899. It seems like that got removed at some point though, but I could not find any discussion around the removal. Was that intentional or was it lost due to something like force-pushing as happened in #775?

I guess this can be considered a feature request for a vdirsyncer module or again a full-blown calendar and contact infrastructure, if that is wanted.

If I overlooked some discussion though I'd be happy to get a link.

**Comment by sumnerevans** [comment]
There was some progress here: https://github.com/nix-community/home-manager/pull/754

I think the biggest thing is just that nobody has had the time to create the full-blown calendar/contact infrastructure; it's a lot of work.

I think that whoever makes the next attempt at building it out should try to do it as iteratively as possible. We don't need to get the whole thing done in one swoop, I am (and I'm sure @berbiche and @rycee are) more than happy to merge in incremental steps that are self-contained but move us in the correct direction.

**Comment by RaitoBezarius** [comment]
FWIW, for vdirsyncer; here's: https://github.com/RaitoBezarius/nixos-home/blob/master/calendar.nix + https://github.com/RaitoBezarius/nixos-home/blob/master/modules/vdirsyncer.nix

**Comment by RaitoBezarius** [comment]
Alas, I do not have any time to upstream them at the moment.

**Comment by Misterio77** [comment]
I'm very interested in getting vdirsyncer, todoman, khard, and khal working. I might give this a shot

**Comment by ambroisie** [comment]
@berbiche shouldn't this be re-opened ?

**Comment by ambroisie** [comment]
Still relevant.

**Comment by jorsn** [comment]
> I think the biggest thing is just that nobody has had the time to create the full-blown calendar/contact infrastructure; it's a lot of work.

I think, the problem back then was not that nobody had the time to create something, but that at some point no authority (with merge access) had the time or interest to give feedback and define what we should do to get something merged, or whether there was any reason not to merge (see [1],[2],[3]). So we ended up just implementing what we were using ourselves and maintaining our personal fork (= fixing the same merge conflict on every update, at Git's fault ([possible solution](https://pijul.org))).

[1]: https://github.com/nix-community/home-manager/pull/754#issuecomment-619348042
[2]: https://github.com/nix-community/home-manager/pull/754#issuecomment-639256037
[3]: https://github.com/nix-community/home-manager/pull/754#issuecomment-694791879

**Comment by teto** [comment]
I am interested in this but was worried about vdirsyncer since at the time it seemed unsupported. Ideally we would have someone invested in this as a HM collaborator (with merge rights) since this is definitely going to be an active part of HM.

**Comment by d-dervishi** [comment]
I'm also interested in this. I'd like to add calendar support to the Thunderbird module, but I'd rather avoid writing Thunderbird-specific options just to remove them once a shared infrastructure is implemented. Is anyone working on this? Are @jorsn's [comments](https://github.com/nix-community/home-manager/issues/2335#issuecomment-1264683111) still relevant?

**Comment by teto** [comment]
seems like we have several candidates for a common architecture: thunderbird, khard, todoman. 
As stated earlier I am willing to review/merge PRs (but can't actively work on a PR).

**Comment by RaitoBezarius** [comment]
I can send vdirsyncer stuff that I have.

**Comment by RaitoBezarius** [comment]
First step https://github.com/nix-community/home-manager/pull/3912

**Comment by teto** [comment]
I've merged the 4/5 year old branch https://github.com/nix-community/home-manager/pull/4078 instead of waiting for the perfect MR (perfect is the enemy of good). This is the beginning of a new cycle so we have 6 months to make the calendar infra "right" (hopefully less). So please start using it and let's not be afraid of refactors/breaking stuff. I've merged because I could configure khal /vdirsyncer with it but ofc there are things I am not happy with it.

**Comment by DanielSiepmann** [comment]
Thanks for bringing this forward @teto. I'll try to migrate and provide feedback.
- Maybe we should trim traililing "/" from basePath. I added a trailing "/" which ended up in wrong paths for khal.
- `warning: unknown key or subsection "highlight_event_days" in section "[calendars][[private]]` comes when executing `khal interactive`. I'm on nixpkgs unstable.
- Looks like `modules/services/vdirsyncer.nix` is not included yet.
- I've multiple accounts with multiple calendars. It looks like this isn't possible with khal yet? As it will add one calendar per account. It uses some `/*` at the end. But opening `khal interactive` will create a folder `*` instead of fetching events from all folders. Looks like this should only be used if type is set to discover. But `*` is always added, ignoring the type. Maybe this is broken right now? I guess the hardcoded `/*` should be removed and the existing code would work picking up globPath with default `*` if set to discover?

**Comment by teto** [comment]
@DanielSiepmann thanks for the feedback

> warning: unknown key or subsection "highlight_event_days" in section "[calendars][[private]]

indeed. Will remove it. I think the key exists but maybe as a global instead. We shouldn't change defaults but I would like to be able to set this option. Its something that escaped my fork.

> Looks like modules/services/vdirsyncer.nix is not included yet.

bad me, I need to double check later.

> I've multiple accounts with multiple calendars. It looks like this isn't possible with khal yet?

I have only one account that is multicalendar and adding `/*` at the end of the caldav sync worked for me but it may have broken the single calendar scenario. I am not sure if I merged that. I need to do another pass. Thanks for feedback.

**Comment by DanielSiepmann** [comment]
@teto feel free to ping me within PRs. I'll try to test them with my setup. I've a separate branch with my migration as WIP.

> I have only one account that is multicalendar and adding `/*` at the end of the caldav sync worked for me but it may have broken the single calendar scenario. I am not sure if I merged that. I need to do another pass. Thanks for feedback.

I needed to set the following in order to make discovering work with current state:
```
          type = "discover";
          glob = "";
```
As this will lead to
```
path = /home/daniels/.local/share/vdir/calendars/private//*
type=discover
```
`/*` needs to be added. But that seems to be the default for `glob` already which is used when type is set to `discover`. But it should not be added if type is not discover. https://github.com/nix-community/home-manager/blob/master/modules/programs/khal.nix#L32 always adds the globbing right now.

**Comment by DanielSiepmann** [comment]
@teto I've created two small PRs. One "fixes" the discover / globbing of khal: https://github.com/nix-community/home-manager/pull/4088 The other adds the missing include: https://github.com/nix-community/home-manager/pull/4089

I was able to fully migrate my existing setup to the new one.

The `highlight_event_days` warning only appears for "calendar" entries. But docs don't mention any limitation and it is only a warning.
